// @poliklot/prettier-plugin-handlebars/dist/cli.js
#!/usr/bin/env node
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.main = main;
exports.parseArgs = parseArgs;
exports.renderHelp = renderHelp;
exports.getDefaultCommand = getDefaultCommand;
const node_path_1 = __importDefault(require("node:path"));
const init_1 = require("./init");
function main(argv, executableName = 'prettier-plugin-handlebars') {
    const options = parseArgs(argv, getDefaultCommand(executableName));
    if (options.help || !options.command) {
        process.stdout.write(renderHelp(executableName));
        return options.help ? 0 : 1;
    }
    if (options.command !== 'init') {
        process.stderr.write(`Unknown command: ${options.command}\n\n`);
        process.stderr.write(renderHelp(executableName));
        return 1;
    }
    const plan = (0, init_1.createInitPlan)(options.cwd);
    if (options.write) {
        (0, init_1.applyInitPlan)(plan);
        process.stdout.write((0, init_1.renderInitPlan)(plan, 'write'));
        return (0, init_1.hasCheckFailures)((0, init_1.createInitPlan)(options.cwd)) ? 1 : 0;
    }
    if (options.check) {
        process.stdout.write((0, init_1.renderInitPlan)(plan, 'check'));
        return (0, init_1.hasCheckFailures)(plan) ? 1 : 0;
    }
    process.stdout.write((0, init_1.renderInitPlan)(plan, 'dry-run'));
    return 0;
}
function parseArgs(argv, defaultCommand) {
    const options = {
        cwd: process.cwd(),
        write: false,
        check: false,
        help: false,
    };
    const args = [...argv];
    while (args.length > 0) {
        const arg = args.shift();
        if (!arg) {
            continue;
        }
        if (arg === '--help' || arg === '-h') {
            options.help = true;
            continue;
        }
        if (arg === '--write') {
            options.write = true;
            continue;
        }
        if (arg === '--check') {
            options.check = true;
            continue;
        }
        if (arg === '--dry-run') {
            options.write = false;
            continue;
        }
        if (arg === '--cwd') {
            const cwd = args.shift();
            if (!cwd) {
                throw new Error('--cwd requires a directory');
            }
            options.cwd = cwd;
            continue;
        }
        if (!options.command) {
            options.command = arg;
            continue;
        }
        throw new Error(`Unexpected argument: ${arg}`);
    }
    if (options.write && options.check) {
        throw new Error('Use either --write or --check, not both.');
    }
    if (!options.command && defaultCommand) {
        options.command = defaultCommand;
    }
    return options;
}
function getDefaultCommand(executableName) {
    return executableName === 'hbs-prettier-init' ? 'init' : undefined;
}
function renderHelp(executableName = 'prettier-plugin-handlebars') {
    const isInitAlias = executableName === 'hbs-prettier-init';
    const usage = isInitAlias
        ? '  hbs-prettier-init [--write|--check] [--cwd <dir>]'
        : '  prettier-plugin-handlebars init [--write|--check] [--cwd <dir>]';
    return [
        '@poliklot/prettier-plugin-handlebars',
        '',
        'Usage:',
        usage,
        '',
        'Commands:',
        '  init        Audit or configure a project for .hbs / .handlebars formatting.',
        '',
        'Options:',
        '  --write    Apply supported config changes. Without this flag, init is a dry run.',
        '  --check    Exit with a non-zero status when setup is incomplete.',
        '  --dry-run  Print the setup report without writing files. This is the default.',
        '  --cwd DIR  Run against another project directory.',
        '  -h, --help Show this help message.',
        '',
    ].join('\n');
}
if (require.main === module) {
    try {
        process.exitCode = main(process.argv.slice(2), node_path_1.default.basename(process.argv[1] ?? ''));
    }
    catch (error) {
        process.stderr.write(`${error instanceof Error ? error.message : String(error)}\n`);
        process.exitCode = 1;
    }
}
