// @poliklot/prettier-plugin-handlebars/dist/init.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handlebarsParserName = exports.pluginPackageName = void 0;
exports.createInitPlan = createInitPlan;
exports.applyInitPlan = applyInitPlan;
exports.hasCheckFailures = hasCheckFailures;
exports.renderInitPlan = renderInitPlan;
const node_fs_1 = __importDefault(require("node:fs"));
const node_path_1 = __importDefault(require("node:path"));
exports.pluginPackageName = '@poliklot/prettier-plugin-handlebars';
exports.handlebarsParserName = 'handlebars';
const defaultConfigFileName = '.prettierrc.json';
const handlebarsOverride = {
    files: ['*.hbs', '*.handlebars'],
    options: {
        parser: exports.handlebarsParserName,
    },
};
const configCandidates = [
    '.prettierrc',
    '.prettierrc.json',
    '.prettierrc.json5',
    '.prettierrc.js',
    '.prettierrc.cjs',
    '.prettierrc.mjs',
    '.prettierrc.yml',
    '.prettierrc.yaml',
    '.prettierrc.toml',
    'prettier.config.js',
    'prettier.config.cjs',
    'prettier.config.mjs',
    'package.json',
];
function createInitPlan(cwd = process.cwd()) {
    const root = node_path_1.default.resolve(cwd);
    const changes = [];
    const warnings = [];
    const installCommand = getInstallCommand(root);
    const locatedConfig = findPrettierConfig(root);
    if (!hasLocalDependency(root, 'prettier')) {
        warnings.push({
            kind: 'dependency',
            relativePath: 'package.json',
            filePath: node_path_1.default.join(root, 'package.json'),
            message: `Install Prettier locally: ${installCommand}`,
            checkFailure: true,
        });
    }
    if (!hasLocalDependency(root, exports.pluginPackageName)) {
        warnings.push({
            kind: 'dependency',
            relativePath: 'package.json',
            filePath: node_path_1.default.join(root, 'package.json'),
            message: `Install the Handlebars plugin locally: ${installCommand}`,
            checkFailure: true,
        });
    }
    if (locatedConfig) {
        if (locatedConfig.kind === 'unsupported') {
            if (!unsupportedConfigLooksConfigured(locatedConfig.filePath)) {
                warnings.push({
                    kind: 'unsupported-config',
                    filePath: locatedConfig.filePath,
                    relativePath: locatedConfig.relativePath,
                    message: `Detected ${locatedConfig.relativePath}. This init command does not rewrite JS/JSON5/YAML/TOML configs automatically. Add ${exports.pluginPackageName} and the *.hbs override manually, or use ${defaultConfigFileName}.`,
                    checkFailure: true,
                });
            }
        }
        else {
            const configChange = planSupportedConfigUpdate(root, locatedConfig);
            warnings.push(...configChange.warnings);
            if (configChange.change) {
                changes.push(configChange.change);
            }
        }
    }
    else {
        const filePath = node_path_1.default.join(root, defaultConfigFileName);
        changes.push({
            kind: 'create',
            filePath,
            relativePath: defaultConfigFileName,
            message: `Create ${defaultConfigFileName} with the plugin and explicit Handlebars parser override.`,
            content: `${JSON.stringify(createRecommendedPrettierConfig(), null, 2)}\n`,
        });
    }
    warnings.push(...findPrettierIgnoreWarnings(root));
    return {
        cwd: root,
        configPath: locatedConfig?.filePath,
        configRelativePath: locatedConfig?.relativePath,
        changes,
        warnings,
        alreadyConfigured: changes.length === 0 && warnings.every((warning) => !warning.checkFailure),
        installCommand,
    };
}
function applyInitPlan(plan) {
    for (const change of plan.changes) {
        node_fs_1.default.writeFileSync(change.filePath, change.content, 'utf8');
    }
}
function hasCheckFailures(plan) {
    return plan.changes.length > 0 || plan.warnings.some((warning) => warning.checkFailure);
}
function renderInitPlan(plan, mode) {
    const lines = [];
    lines.push('@poliklot/prettier-plugin-handlebars init');
    lines.push(`Project: ${plan.cwd}`);
    lines.push(`Mode: ${mode}`);
    lines.push('');
    if (plan.configRelativePath) {
        lines.push(`Detected Prettier config: ${plan.configRelativePath}`);
    }
    else {
        lines.push(`Detected Prettier config: none`);
    }
    lines.push('');
    if (plan.changes.length === 0) {
        lines.push('Config changes: none');
    }
    else {
        lines.push('Config changes:');
        for (const change of plan.changes) {
            lines.push(`- ${change.kind.toUpperCase()} ${change.relativePath}: ${change.message}`);
        }
    }
    if (plan.warnings.length > 0) {
        lines.push('');
        lines.push('Warnings:');
        for (const warning of plan.warnings) {
            const location = warning.relativePath ? `${warning.relativePath}: ` : '';
            lines.push(`- ${location}${warning.message}`);
        }
    }
    lines.push('');
    if (mode === 'dry-run') {
        if (plan.changes.length > 0) {
            lines.push('Run with --write to apply these config changes.');
        }
        else {
            lines.push('No config writes are needed.');
        }
    }
    else if (mode === 'write') {
        lines.push(plan.changes.length > 0 ? 'Applied config changes.' : 'No config writes were needed.');
    }
    else if (mode === 'check') {
        lines.push(hasCheckFailures(plan) ? 'Check failed: setup is incomplete.' : 'Check passed: setup looks ready.');
    }
    if (plan.installCommand && plan.warnings.some((warning) => warning.kind === 'dependency')) {
        lines.push(`Install command: ${plan.installCommand}`);
    }
    return `${lines.join('\n')}\n`;
}
function findPrettierConfig(cwd) {
    for (const candidate of configCandidates) {
        const filePath = node_path_1.default.join(cwd, candidate);
        if (!node_fs_1.default.existsSync(filePath)) {
            continue;
        }
        if (candidate === 'package.json') {
            const packageJson = readJsonFile(filePath);
            if (packageJson && isRecord(packageJson) && isRecord(packageJson.prettier)) {
                return {
                    filePath,
                    relativePath: candidate,
                    kind: 'package-json',
                };
            }
            continue;
        }
        if (candidate === '.prettierrc' || candidate.endsWith('.json')) {
            return {
                filePath,
                relativePath: candidate,
                kind: 'json',
            };
        }
        return {
            filePath,
            relativePath: candidate,
            kind: 'unsupported',
        };
    }
    return undefined;
}
function planSupportedConfigUpdate(root, locatedConfig) {
    const warnings = [];
    const source = node_fs_1.default.readFileSync(locatedConfig.filePath, 'utf8');
    const parsed = parseJson(source, locatedConfig.filePath);
    if (!isRecord(parsed)) {
        return {
            warnings: [
                {
                    kind: 'invalid-config',
                    filePath: locatedConfig.filePath,
                    relativePath: locatedConfig.relativePath,
                    message: `Could not parse ${locatedConfig.relativePath} as a JSON object. Add the plugin config manually.`,
                    checkFailure: true,
                },
            ],
        };
    }
    if (locatedConfig.kind === 'package-json') {
        const packageJson = cloneRecord(parsed);
        const prettierConfig = isRecord(packageJson.prettier) ? cloneRecord(packageJson.prettier) : {};
        const ensured = ensureRecommendedConfig(prettierConfig, root, locatedConfig.relativePath);
        warnings.push(...ensured.warnings);
        if (!ensured.changed) {
            return { warnings };
        }
        packageJson.prettier = ensured.value;
        return {
            warnings,
            change: {
                kind: 'update',
                filePath: locatedConfig.filePath,
                relativePath: locatedConfig.relativePath,
                message: ensured.messages.join(' '),
                content: `${JSON.stringify(packageJson, null, 2)}\n`,
            },
        };
    }
    const config = cloneRecord(parsed);
    const ensured = ensureRecommendedConfig(config, root, locatedConfig.relativePath);
    warnings.push(...ensured.warnings);
    if (!ensured.changed) {
        return { warnings };
    }
    return {
        warnings,
        change: {
            kind: 'update',
            filePath: locatedConfig.filePath,
            relativePath: locatedConfig.relativePath,
            message: ensured.messages.join(' '),
            content: `${JSON.stringify(ensured.value, null, 2)}\n`,
        },
    };
}
function ensureRecommendedConfig(config, cwd, relativePath) {
    const value = cloneRecord(config);
    const messages = [];
    const warnings = [];
    let changed = false;
    if (value.plugins === undefined) {
        value.plugins = [exports.pluginPackageName];
        messages.push(`Added ${exports.pluginPackageName} to plugins.`);
        changed = true;
    }
    else if (Array.isArray(value.plugins)) {
        if (!value.plugins.includes(exports.pluginPackageName)) {
            value.plugins = [...value.plugins, exports.pluginPackageName];
            messages.push(`Added ${exports.pluginPackageName} to plugins.`);
            changed = true;
        }
    }
    else {
        warnings.push({
            kind: 'invalid-config',
            relativePath,
            filePath: node_path_1.default.join(cwd, relativePath),
            message: '`plugins` exists but is not an array. Add the Handlebars plugin manually.',
            checkFailure: true,
        });
    }
    if (value.overrides === undefined) {
        value.overrides = [handlebarsOverride];
        messages.push('Added explicit *.hbs / *.handlebars parser override.');
        changed = true;
    }
    else if (Array.isArray(value.overrides)) {
        const overrideResult = ensureHandlebarsOverride(value.overrides);
        if (overrideResult.changed) {
            value.overrides = overrideResult.overrides;
            messages.push(overrideResult.message);
            changed = true;
        }
        warnings.push(...overrideResult.warnings.map((warning) => ({ ...warning, filePath: node_path_1.default.join(cwd, relativePath), relativePath })));
    }
    else {
        warnings.push({
            kind: 'invalid-config',
            relativePath,
            filePath: node_path_1.default.join(cwd, relativePath),
            message: '`overrides` exists but is not an array. Add the Handlebars parser override manually.',
            checkFailure: true,
        });
    }
    return {
        value,
        changed,
        messages,
        warnings,
    };
}
function ensureHandlebarsOverride(overrides) {
    const nextOverrides = overrides.map((override) => (isRecord(override) ? cloneRecord(override) : override));
    const warnings = [];
    let changed = false;
    let message = '';
    const matchingIndex = nextOverrides.findIndex((override) => isRecord(override) && fileSpecContainsHandlebars(override.files));
    if (matchingIndex === -1) {
        nextOverrides.push(handlebarsOverride);
        return {
            overrides: nextOverrides,
            changed: true,
            message: 'Added explicit *.hbs / *.handlebars parser override.',
            warnings,
        };
    }
    const matchingOverride = nextOverrides[matchingIndex];
    if (!isRecord(matchingOverride)) {
        return {
            overrides: nextOverrides,
            changed,
            message,
            warnings,
        };
    }
    const options = isRecord(matchingOverride.options) ? cloneRecord(matchingOverride.options) : {};
    if (options.parser === exports.handlebarsParserName) {
        return {
            overrides: nextOverrides,
            changed,
            message,
            warnings,
        };
    }
    if (options.parser === undefined) {
        options.parser = exports.handlebarsParserName;
        matchingOverride.options = options;
        nextOverrides[matchingIndex] = matchingOverride;
        changed = true;
        message = 'Added parser: "handlebars" to the existing Handlebars override.';
    }
    else {
        nextOverrides.push(handlebarsOverride);
        changed = true;
        message = 'Added a later explicit Handlebars override because an existing *.hbs override uses another parser.';
        warnings.push({
            kind: 'invalid-config',
            message: `An existing *.hbs override uses parser ${JSON.stringify(options.parser)}. The init command appended a later handlebars override so Prettier resolves *.hbs files correctly.`,
            checkFailure: false,
        });
    }
    return {
        overrides: nextOverrides,
        changed,
        message,
        warnings,
    };
}
function createRecommendedPrettierConfig() {
    return {
        plugins: [exports.pluginPackageName],
        overrides: [handlebarsOverride],
    };
}
function findPrettierIgnoreWarnings(cwd) {
    const filePath = node_path_1.default.join(cwd, '.prettierignore');
    if (!node_fs_1.default.existsSync(filePath)) {
        return [];
    }
    const source = node_fs_1.default.readFileSync(filePath, 'utf8');
    const warnings = [];
    const lines = source.split(/\r?\n/);
    lines.forEach((line, index) => {
        const trimmed = line.trim();
        if (!trimmed || trimmed.startsWith('#') || trimmed.startsWith('!')) {
            return;
        }
        if (ignorePatternContainsHandlebars(trimmed)) {
            warnings.push({
                kind: 'prettierignore',
                filePath,
                relativePath: '.prettierignore',
                message: `Line ${index + 1} may skip Handlebars files: ${trimmed}`,
                checkFailure: true,
            });
        }
    });
    return warnings;
}
function ignorePatternContainsHandlebars(pattern) {
    const normalized = pattern.replace(/\\/g, '/').toLowerCase();
    return normalized.includes('.hbs') || normalized.includes('*.hbs') || normalized.includes('handlebars');
}
function unsupportedConfigLooksConfigured(filePath) {
    const source = node_fs_1.default.readFileSync(filePath, 'utf8').toLowerCase();
    return (source.includes(exports.pluginPackageName.toLowerCase()) &&
        source.includes(exports.handlebarsParserName) &&
        (source.includes('.hbs') || source.includes('*.hbs') || source.includes('handlebars')));
}
function fileSpecContainsHandlebars(files) {
    if (typeof files === 'string') {
        return files.toLowerCase().includes('.hbs') || files.toLowerCase().includes('handlebars');
    }
    if (Array.isArray(files)) {
        return files.some((file) => fileSpecContainsHandlebars(file));
    }
    return false;
}
function hasLocalDependency(cwd, dependencyName) {
    const packagePath = node_path_1.default.join(cwd, 'package.json');
    const packageJson = readJsonFile(packagePath);
    if (!isRecord(packageJson)) {
        return false;
    }
    return [
        packageJson.dependencies,
        packageJson.devDependencies,
        packageJson.peerDependencies,
        packageJson.optionalDependencies,
    ].some((dependencies) => isRecord(dependencies) && typeof dependencies[dependencyName] === 'string');
}
function getInstallCommand(cwd) {
    if (node_fs_1.default.existsSync(node_path_1.default.join(cwd, 'pnpm-lock.yaml'))) {
        return `pnpm add -D prettier ${exports.pluginPackageName}`;
    }
    if (node_fs_1.default.existsSync(node_path_1.default.join(cwd, 'yarn.lock'))) {
        return `yarn add -D prettier ${exports.pluginPackageName}`;
    }
    if (node_fs_1.default.existsSync(node_path_1.default.join(cwd, 'bun.lockb')) || node_fs_1.default.existsSync(node_path_1.default.join(cwd, 'bun.lock'))) {
        return `bun add -d prettier ${exports.pluginPackageName}`;
    }
    return `npm install --save-dev prettier ${exports.pluginPackageName}`;
}
function readJsonFile(filePath) {
    if (!node_fs_1.default.existsSync(filePath)) {
        return undefined;
    }
    return parseJson(node_fs_1.default.readFileSync(filePath, 'utf8'), filePath);
}
function parseJson(source, filePath) {
    try {
        return JSON.parse(source);
    }
    catch {
        return undefined;
    }
}
function cloneRecord(record) {
    return { ...record };
}
function isRecord(value) {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
}
