// @poliklot/prettier-plugin-handlebars/dist/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultOptions = exports.options = exports.printers = exports.parsers = exports.languages = void 0;
const parser_1 = require("./parser");
const printer_1 = require("./printer");
exports.languages = [
    {
        name: 'Handlebars',
        type: 'markup',
        parsers: ['handlebars'],
        extensions: ['.hbs', '.handlebars'],
        aliases: ['hbs', 'htmlbars', 'classic-handlebars'],
        vscodeLanguageIds: ['handlebars'],
    },
];
exports.parsers = {
    'handlebars': {
        parse: parser_1.parse,
        astFormat: 'handlebars-ast',
        locStart: parser_1.locStart,
        locEnd: parser_1.locEnd,
    },
};
exports.printers = {
    'handlebars-ast': printer_1.printer,
};
exports.options = {
    dataAttributeOrder: {
        since: '0.1.0',
        category: 'HTML',
        type: 'string',
        array: true,
        default: [{ value: [] }],
        description: 'Ordering override for data-* attributes.',
    },
    maxEmptyLines: {
        since: '0.1.0',
        category: 'HTML',
        type: 'int',
        default: 1,
        description: 'Maximum number of consecutive blank lines to preserve between nodes.',
    },
};
exports.defaultOptions = {};
