// @poliklot/prettier-plugin-handlebars/dist/printer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.printer = void 0;
const doc_1 = require("prettier/doc");
const template_format_core_1 = require("template-format-core");
const template_format_core_2 = require("template-format-core");
const tokens_1 = require("./dialects/handlebars/tokens");
const { hardline, join, group, indent, align, line, softline, ifBreak, lineSuffix, lineSuffixBoundary } = doc_1.builders;
const { stripTrailingHardline, willBreak } = doc_1.utils;
const mapDoc = doc_1.utils.mapDoc;
const concat = doc_1.builders.concat;
const templateDialect = tokens_1.handlebarsDialect;
function docHasHardline(doc) {
    if (typeof doc === 'string') {
        return doc.includes('\n');
    }
    if (typeof doc === 'number' || doc === null || doc === undefined) {
        return false;
    }
    if (doc === hardline) {
        return true;
    }
    if (Array.isArray(doc)) {
        return doc.some(docHasHardline);
    }
    if (typeof doc === 'object' && 'contents' in doc) {
        return docHasHardline(doc.contents);
    }
    if (typeof doc === 'object' && 'parts' in doc) {
        return docHasHardline(doc.parts);
    }
    return false;
}
function docBreaks(doc) {
    if (Array.isArray(doc)) {
        return doc.some(docBreaks);
    }
    if (typeof doc === 'object' && doc !== null && 'contents' in doc) {
        return docBreaks(doc.contents);
    }
    if (typeof doc === 'object' && doc !== null && 'parts' in doc) {
        return docBreaks(doc.parts);
    }
    return docHasHardline(doc) || willBreak(doc);
}
function getTrimOpen(node) {
    return node.trimOpen ? '~' : '';
}
function getTrimClose(node) {
    return node.trimClose ? '~' : '';
}
function getTemplateTagDelimiters(triple = false) {
    return templateDialect.getTagDelimiters(triple);
}
function buildTemplateTag(content, trimOpen = '', trimClose = '', triple = false) {
    const { open, close } = getTemplateTagDelimiters(triple);
    return `${open}${trimOpen}${content}${trimClose}${close}`;
}
function isSimpleValueMustache(node) {
    return node.params.length === 0 && node.hash.length === 0 && (!node.blockParams || node.blockParams.length === 0);
}
function getMustacheOpenPadding(node, content) {
    return content.length > 0 && isSimpleValueMustache(node) ? ' ' : '';
}
function getMustacheClosePadding(node, content) {
    if (content.length > 0 && isSimpleValueMustache(node)) {
        return ' ';
    }
    return node.trimClose && /\s/.test(content) ? ' ' : '';
}
function getTrimClosePadding(node, content) {
    return node.trimClose && /\s/.test(content) ? ' ' : '';
}
function shouldKeepParamInline(param) {
    return param.includes('\n') || /^\(parseJSON\s+['"`]/.test(param.trim());
}
function getBlockPrefix(node) {
    return node.blockPrefix ?? '#';
}
function hasInlineBoundaryWhitespace(value) {
    return typeof value === 'string' && /\s/.test(value);
}
function isPunctuationOnlyTextNode(node) {
    return node?.type === 'TextNode' && /^[.,:;!?+"'«»+]+$/.test(node.value);
}
function isPlainAttribute(attr) {
    return attr.type === 'Attribute';
}
function isRawAttribute(attr) {
    return attr.type === 'RawAttribute';
}
function getMaxEmptyLines(options) {
    const rawValue = options.maxEmptyLines;
    if (typeof rawValue === 'number' && rawValue >= 0) {
        return rawValue;
    }
    return 1;
}
exports.printer = {
    getVisitorKeys(node, nonTraversableKeys) {
        return getHandlebarsVisitorKeys(node, nonTraversableKeys);
    },
    hasPrettierIgnore(path) {
        const node = path.getValue();
        return node?.type === 'CommentStatement' && hasCommentDirective(node, 'prettier-ignore');
    },
    canAttachComment(node) {
        return node.type !== 'CommentStatement' && node.type !== 'UnmatchedNode';
    },
    isBlockComment(node) {
        return node.type === 'CommentStatement' && (node.block || node.multiline);
    },
    willPrintOwnComments(path) {
        return path.getValue()?.type === 'CommentStatement';
    },
    printComment(path, options) {
        return printCommentStatement(path.getValue(), options);
    },
    getCommentChildNodes(node) {
        return getCommentChildNodes(node);
    },
    embed(path, options) {
        const node = path.getValue();
        if (node.type !== 'TextNode') {
            return null;
        }
        const parentNode = path.getParentNode();
        if (parentNode?.type !== 'ElementNode') {
            return null;
        }
        const parser = getEmbeddedRawTextParser(parentNode, node, options);
        if (!parser) {
            return null;
        }
        return async (textToDoc) => {
            const content = normalizeEmbeddedRawText(node.value);
            if (content.trim() === '') {
                return '';
            }
            const prepared = prepareEmbeddedRawText(content, parser);
            if (!prepared) {
                return formatVerbatimText(content);
            }
            try {
                const doc = await textToDoc(prepared.text, {
                    ...options,
                    parser,
                });
                return stripTrailingHardline(restoreHandlebarsPlaceholders(doc, prepared.replacements));
            }
            catch {
                return formatVerbatimText(content);
            }
        };
    },
    print(path, options, print) {
        const node = path.getValue();
        switch (node.type) {
            case 'Program':
                return printProgram(path, options, print);
            case 'ElementNode':
                return printElement(path, options, print);
            case 'TextNode':
                if (node.verbatim) {
                    if (node.preserveWhitespace) {
                        return node.value;
                    }
                    const parentNode = path.getParentNode();
                    const value = shouldTrimRawTextBoundaryWhitespace(parentNode, node)
                        ? trimRawTextBoundaryWhitespace(node.value)
                        : node.value;
                    return formatVerbatimText(value);
                }
                if (node.blankLines) {
                    const maxEmptyLines = getMaxEmptyLines(options);
                    const allowedBlankLines = Math.min(node.blankLines, maxEmptyLines);
                    const extraHardlines = allowedBlankLines - 1;
                    return extraHardlines > 0 ? concat(new Array(extraHardlines).fill(hardline)) : '';
                }
                return node.value.replace(/\s+/g, ' ').trim();
            case 'MustacheStatement':
                return printMustache(node, options);
            case 'DecoratorStatement':
                return printDecorator(node, options);
            case 'BlockStatement':
                return printBlock(path, options, print);
            case 'PartialStatement':
                return printPartial(node, options);
            case 'CommentStatement':
                return printCommentStatement(node, options);
            case 'UnmatchedNode':
                return node.raw;
            default:
                return '';
        }
    },
};
function getHandlebarsVisitorKeys(node, nonTraversableKeys) {
    const type = node?.type;
    return getNodeVisitorKeys(type).filter((key) => !nonTraversableKeys.has(key));
}
function getNodeVisitorKeys(type) {
    switch (type) {
        case 'Program':
            return ['body'];
        case 'ElementNode':
            return ['attributes', 'children'];
        case 'Attribute':
            return ['value'];
        case 'AttributeValue':
            return ['parts'];
        case 'AttributeBlock':
            return ['block'];
        case 'BlockStatement':
            return ['program', 'inverseChain', 'inverse'];
        case 'ElseBranch':
            return ['program'];
        default:
            return [];
    }
}
function getCommentChildNodes(node) {
    switch (node.type) {
        case 'Program':
            return node.body;
        case 'ElementNode': {
            const element = node;
            const attributeNodes = element.attributes.flatMap((attr) => {
                if (attr.type === 'AttributeBlock') {
                    return [attr.block];
                }
                if (attr.type === 'Attribute' && attr.value) {
                    return attr.value.parts;
                }
                return [];
            });
            return [...attributeNodes, ...element.children];
        }
        case 'BlockStatement': {
            const block = node;
            return [
                ...block.program.body,
                ...(block.inverseChain ?? []).flatMap((branch) => branch.program.body),
                ...block.inverse.body,
            ];
        }
        default:
            return [];
    }
}
function hasCommentDirective(node, directive) {
    return node.value.toLowerCase().includes(directive);
}
function printCommentStatement(node, options) {
    if (node.multiline) {
        return formatMultilineComment(node.value, options, node.inline);
    }
    if (!node.block && node.value.startsWith('<')) {
        return templateDialect.getLineCommentTag(node.value);
    }
    if (node.block) {
        const trimmedValue = typeof node.value === 'string' ? node.value.replace(/[ \t]+$/gm, '') : node.value;
        return templateDialect.getBlockCommentTag(trimmedValue);
    }
    return templateDialect.getLineCommentTag(node.value);
}
function formatVerbatimText(content) {
    const withoutLeadingNewline = content.startsWith('\n') ? content.slice(1) : content;
    const withoutTrailingNewline = withoutLeadingNewline.endsWith('\n')
        ? withoutLeadingNewline.slice(0, -1)
        : withoutLeadingNewline;
    if (withoutTrailingNewline.trimStart().startsWith('<!--')) {
        return withoutTrailingNewline;
    }
    const lines = withoutTrailingNewline.split('\n');
    const commonIndent = lines.reduce((min, line) => {
        if (line.trim() === '')
            return min;
        const indentLength = (line.match(/^[ \t]*/) || [''])[0].length;
        return Math.min(min, indentLength);
    }, Number.MAX_SAFE_INTEGER);
    const normalizedIndent = Number.isFinite(commonIndent) ? commonIndent : 0;
    let normalizedLines = lines.map((line) => {
        const indentLength = (line.match(/^[ \t]*/) || [''])[0].length;
        return line.slice(Math.min(indentLength, normalizedIndent));
    });
    while (normalizedLines.length > 0 && normalizedLines[0].trim() === '') {
        normalizedLines = normalizedLines.slice(1);
    }
    while (normalizedLines.length > 0 && normalizedLines[normalizedLines.length - 1].trim() === '') {
        normalizedLines = normalizedLines.slice(0, -1);
    }
    if (normalizedLines.length === 0) {
        return '';
    }
    const docs = [];
    normalizedLines.forEach((lineText, index) => {
        const trailingWhitespaceMatch = lineText.match(/(\s+)$/);
        const trailingWhitespace = trailingWhitespaceMatch?.[1] ?? '';
        const contentWithoutTrailing = trailingWhitespace ? lineText.slice(0, -trailingWhitespace.length) : lineText;
        docs.push(contentWithoutTrailing);
        if (trailingWhitespace) {
            docs.push(lineSuffix(trailingWhitespace));
            docs.push(lineSuffixBoundary);
        }
        if (index < normalizedLines.length - 1) {
            docs.push(hardline);
        }
    });
    return concat(docs);
}
function shouldTrimRawTextBoundaryWhitespace(parentNode, node) {
    return (parentNode?.type === 'ElementNode' &&
        template_format_core_1.trimmableRawTextElements.has(parentNode.tag.toLowerCase()) &&
        node.verbatim === true &&
        node.preserveWhitespace !== true);
}
function trimRawTextBoundaryWhitespace(value) {
    return value.replace(/[ \t]+$/, '');
}
function getEmbeddedRawTextParser(element, child, options) {
    if (!isEmbeddedLanguageFormattingEnabled(options) || !isSingleRawTextChild(element, child)) {
        return null;
    }
    const tag = element.tag.toLowerCase();
    const content = normalizeEmbeddedRawText(child.value);
    let parser = null;
    if (tag === 'style') {
        const type = getStaticAttributeValue(element, 'type');
        parser = !type || type === 'text/css' ? 'css' : null;
    }
    else if (tag === 'script') {
        if (hasPlainAttribute(element, 'src')) {
            return null;
        }
        const type = getStaticAttributeValue(element, 'type');
        parser = isJavaScriptScriptType(type) ? 'babel' : null;
    }
    if (!parser || !canFormatEmbeddedRawText(content, tag, parser)) {
        return null;
    }
    return parser;
}
function isEmbeddedLanguageFormattingEnabled(options) {
    return options.embeddedLanguageFormatting !== 'off';
}
function isSingleRawTextChild(element, child) {
    return (template_format_core_1.trimmableRawTextElements.has(element.tag.toLowerCase()) &&
        element.children.length === 1 &&
        element.children[0] === child &&
        child.verbatim === true);
}
function canFormatEmbeddedRawText(content, tag, parser) {
    return (content.trim() !== '' &&
        !new RegExp(`</\\s*${tag}`, 'i').test(content) &&
        !(tag === 'style' && hasMultilineBlockComment(content)) &&
        prepareEmbeddedRawText(content, parser) !== null);
}
function prepareEmbeddedRawText(content, parser) {
    const tokens = findEmbeddedHandlebarsTokens(content);
    if (tokens.length === 0) {
        return { text: content, replacements: new Map() };
    }
    if (tokens.some((token) => isUnsafeEmbeddedHandlebarsToken(content, token))) {
        return null;
    }
    const replacements = new Map();
    let prepared = '';
    let lastIndex = 0;
    tokens.forEach((token, index) => {
        const placeholder = parser === 'css' ? `poliklot-hbs-placeholder-${index}` : `__POLIKLOT_HBS_PLACEHOLDER_${index}__`;
        prepared += content.slice(lastIndex, token.start);
        prepared += placeholder;
        replacements.set(placeholder, content.slice(token.start, token.end));
        lastIndex = token.end;
    });
    prepared += content.slice(lastIndex);
    return { text: prepared, replacements };
}
function restoreHandlebarsPlaceholders(doc, replacements) {
    if (replacements.size === 0) {
        return doc;
    }
    const placeholders = [...replacements.keys()].sort((left, right) => right.length - left.length);
    return mapDoc(doc, (currentDoc) => {
        if (typeof currentDoc !== 'string') {
            return currentDoc;
        }
        return placeholders.reduce((value, placeholder) => {
            const replacement = replacements.get(placeholder) ?? placeholder;
            return value.split(placeholder).join(replacement);
        }, currentDoc);
    });
}
function findEmbeddedHandlebarsTokens(content) {
    const tokens = [];
    let position = 0;
    while (position < content.length) {
        const start = content.indexOf('{{', position);
        if (start === -1) {
            break;
        }
        if (isEscapedEmbeddedHandlebarsOpen(content, start)) {
            position = start + 2;
            continue;
        }
        if (content.startsWith('{{{{', start)) {
            const close = content.indexOf('}}}}', start + 4);
            if (close === -1) {
                return [{ start, end: content.length }];
            }
            tokens.push({ start, end: close + 4 });
            position = close + 4;
            continue;
        }
        const triple = content.startsWith('{{{', start);
        const closeDelimiter = triple ? '}}}' : '}}';
        const close = content.indexOf(closeDelimiter, start + (triple ? 3 : 2));
        if (close === -1) {
            return [{ start, end: content.length }];
        }
        tokens.push({ start, end: close + closeDelimiter.length });
        position = close + closeDelimiter.length;
    }
    return tokens;
}
function isEscapedEmbeddedHandlebarsOpen(content, position) {
    let slashCount = 0;
    for (let index = position - 1; index >= 0 && content[index] === '\\'; index -= 1) {
        slashCount += 1;
    }
    return slashCount % 2 === 1;
}
function isUnsafeEmbeddedHandlebarsToken(content, token) {
    const raw = content.slice(token.start, token.end);
    if (raw.startsWith('{{{{')) {
        return true;
    }
    const triple = raw.startsWith('{{{');
    if ((triple && !raw.endsWith('}}}')) || (!triple && !raw.endsWith('}}'))) {
        return true;
    }
    const openLength = triple ? 3 : 2;
    const closeLength = triple ? 3 : 2;
    const inner = raw.slice(openLength, -closeLength).trim().replace(/^~/, '').replace(/~$/, '').trim();
    return (inner === '' ||
        inner === 'else' ||
        inner.startsWith('else ') ||
        /^[#/!>*<$]/.test(inner) ||
        isStandaloneEmbeddedHandlebarsToken(content, token));
}
function isStandaloneEmbeddedHandlebarsToken(content, token) {
    const lineStart = content.lastIndexOf('\n', token.start - 1) + 1;
    const nextLineBreak = content.indexOf('\n', token.end);
    const lineEnd = nextLineBreak === -1 ? content.length : nextLineBreak;
    return content.slice(lineStart, token.start).trim() === '' && content.slice(token.end, lineEnd).trim() === '';
}
function hasMultilineBlockComment(content) {
    return /\/\*[\s\S]*?\n[\s\S]*?\*\//.test(content);
}
function getStaticAttributeValue(element, name) {
    const attr = element.attributes.find((candidate) => candidate.type === 'Attribute' && candidate.name.toLowerCase() === name);
    if (!attr || attr.type !== 'Attribute') {
        return null;
    }
    if (!attr.value) {
        return '';
    }
    if (!attr.value.parts.every((part) => part.type === 'TextNode')) {
        return null;
    }
    return attr.value.parts.map((part) => part.value).join('').trim().toLowerCase();
}
function hasPlainAttribute(element, name) {
    return element.attributes.some((attr) => attr.type === 'Attribute' && attr.name.toLowerCase() === name);
}
function isJavaScriptScriptType(type) {
    return (!type ||
        type === 'module' ||
        type === 'text/javascript' ||
        type === 'application/javascript' ||
        type === 'text/ecmascript' ||
        type === 'application/ecmascript');
}
function normalizeEmbeddedRawText(content) {
    const lines = (0, template_format_core_2.trimSurroundingBlankLines)(content.replace(/\r\n?/g, '\n').replace(/[ \t]+$/gm, '').split('\n'));
    return (0, template_format_core_2.stripCommonIndent)(lines).join('\n');
}
function printProgram(path, options, print) {
    const parts = [];
    const nodes = [];
    const isRootProgram = !path.getParentNode();
    let rootFragmentDepth = 0;
    path.each((childPath) => {
        const childNode = childPath.getValue();
        if (childNode.type === 'TextNode' && childNode.blankLines && getMaxEmptyLines(options) === 0) {
            return;
        }
        const closingTagName = getHtmlClosingTagName(childNode);
        if (closingTagName && closingTagName !== 'html') {
            rootFragmentDepth = Math.max(rootFragmentDepth - 1, 0);
        }
        let doc = print(childPath);
        if (doc === null) {
            return;
        }
        if (isRootProgram && rootFragmentDepth > 0) {
            doc = applyRootFragmentIndent(doc, rootFragmentDepth, options);
        }
        else if (isRootProgram && shouldPreserveRootClosingTagIndent(childNode)) {
            const standaloneIndent = getOriginalStandaloneIndent(childNode, options);
            if (standaloneIndent) {
                doc = concat([standaloneIndent, doc]);
            }
        }
        else if (isRootProgram && shouldPreserveRootHandlebarsIndent(childNode)) {
            const standaloneIndent = getOriginalStandaloneIndent(childNode, options);
            if (standaloneIndent && !docBreaks(doc)) {
                doc = concat([standaloneIndent, doc]);
            }
        }
        nodes.push(childNode);
        parts.push(doc);
        const openingTagName = getHtmlOpeningTagName(childNode);
        if (openingTagName &&
            openingTagName !== 'html' &&
            !template_format_core_1.voidElements.has(openingTagName) &&
            (rootFragmentDepth > 0 || openingTagName === 'head' || openingTagName === 'body')) {
            rootFragmentDepth += 1;
        }
    }, 'body');
    if (parts.length === 0) {
        return '';
    }
    while (parts.length > 0 && parts[0] === '' && nodes[0]?.type === 'TextNode' && nodes[0].blankLines) {
        parts.shift();
        nodes.shift();
    }
    while (parts.length > 0 &&
        parts[parts.length - 1] === '' &&
        nodes[nodes.length - 1]?.type === 'TextNode' &&
        nodes[nodes.length - 1].blankLines) {
        parts.pop();
        nodes.pop();
    }
    if (parts.length === 0) {
        return '';
    }
    const lastNode = nodes[nodes.length - 1];
    const lastPart = parts[parts.length - 1];
    if (lastNode?.type === 'UnmatchedNode' && typeof lastPart === 'string') {
        parts[parts.length - 1] = lastPart.replace(/\n+$/, '');
    }
    if (canPrintRootInlineTextTemplate(nodes, options)) {
        return concat([stringifyInlineChildren(nodes, options), hardline]);
    }
    return concat([join(hardline, parts), hardline]);
}
function getOriginalStandaloneIndent(node, options) {
    const range = node.range;
    const originalText = options.originalText;
    if (!range || !originalText) {
        return '';
    }
    const nodeText = originalText.slice(range[0], range[1]);
    const nodeSpansMultipleLines = /[\r\n]/.test(nodeText);
    if (nodeSpansMultipleLines && node.type !== 'ElementNode') {
        return '';
    }
    const lineStart = originalText.lastIndexOf('\n', range[0] - 1) + 1;
    const before = originalText.slice(lineStart, range[0]);
    if (!/^[ \t]+$/.test(before)) {
        return '';
    }
    if (!nodeSpansMultipleLines) {
        const nextLineBreak = originalText.indexOf('\n', range[1]);
        const lineEnd = nextLineBreak === -1 ? originalText.length : nextLineBreak;
        const after = originalText.slice(range[1], lineEnd);
        if (!/^[ \t\r]*$/.test(after)) {
            return '';
        }
    }
    return before;
}
function applyRootFragmentIndent(doc, depth, options) {
    const prefix = getIndentUnit(options).repeat(depth);
    return prefix ? concat([prefix, align(prefix, doc)]) : doc;
}
function shouldPreserveRootClosingTagIndent(node) {
    return Boolean(getHtmlClosingTagName(node));
}
function shouldPreserveRootHandlebarsIndent(node) {
    return (node.type === 'MustacheStatement' ||
        node.type === 'PartialStatement' ||
        node.type === 'DecoratorStatement' ||
        node.type === 'BlockStatement');
}
function getUnmatchedRaw(node) {
    return node.type === 'UnmatchedNode' ? (node.raw ?? '').trim() : '';
}
function getTextValue(node) {
    return node.type === 'TextNode' ? (node.value ?? '').trim() : '';
}
function getHtmlOpeningTagName(node) {
    const raw = getUnmatchedRaw(node);
    const match = raw.match(/^<([A-Za-z][\w:-]*)(?:\s|>|\/>)/u);
    return match ? match[1].toLowerCase() : null;
}
function getHtmlClosingTagName(node) {
    const value = getUnmatchedRaw(node) || getTextValue(node);
    const match = value.match(/^<\/([A-Za-z][\w:-]*)\s*>$/u);
    return match ? match[1].toLowerCase() : null;
}
function canPrintRootInlineTextTemplate(nodes, options) {
    return (nodes.some((node) => node.type === 'TextNode') &&
        nodes.every((node, index) => isRootInlineTextTemplateChild(node, index, nodes, options)));
}
function isRootInlineTextTemplateChild(node, index, nodes, options) {
    if (node.type === 'MustacheStatement' || node.type === 'PartialStatement' || node.type === 'DecoratorStatement') {
        return true;
    }
    if (node.type === 'BlockStatement') {
        return canInlineBlock(node, options, 'Program');
    }
    if (node.type === 'CommentStatement') {
        const comment = node;
        return !comment.block && !comment.multiline;
    }
    if (node.type !== 'TextNode') {
        return false;
    }
    const text = node;
    if (text.verbatim || text.blankLines || /[\r\n]/.test(text.value) || hasLineBreak(text.leadingWhitespace)) {
        return false;
    }
    return !hasLineBreak(text.trailingWhitespace) || index === nodes.length - 1;
}
function hasLineBreak(value) {
    return typeof value === 'string' && /[\r\n]/.test(value);
}
function sortAttributes(attributes, options) {
    const sorted = [];
    let buffer = [];
    const flush = () => {
        if (buffer.length === 0)
            return;
        sorted.push(...sortPlainAttributes(buffer, options));
        buffer = [];
    };
    attributes.forEach((attr) => {
        if (!isPlainAttribute(attr)) {
            flush();
            sorted.push(attr);
            return;
        }
        buffer.push(attr);
    });
    flush();
    return sorted;
}
function sortPlainAttributes(attributes, options) {
    const plainAttributes = attributes.filter(isPlainAttribute);
    const others = plainAttributes.filter((attr) => attr.name !== 'id' && attr.name !== 'class');
    const idAttr = plainAttributes.find((attr) => attr.name === 'id');
    const classAttr = plainAttributes.find((attr) => attr.name === 'class');
    const ordered = [];
    if (idAttr)
        ordered.push(idAttr);
    if (classAttr)
        ordered.push(classAttr);
    const preferredDataOrder = options.dataAttributeOrder;
    const dataOrder = Array.isArray(preferredDataOrder) ? preferredDataOrder : [];
    if (dataOrder.length === 0) {
        return ordered.concat(others);
    }
    const orderMap = new Map(dataOrder.map((name, index) => [name, index]));
    const nonDataAttrs = others.filter((attr) => !attr.name.startsWith('data-'));
    const dataAttrs = others.filter((attr) => attr.name.startsWith('data-'));
    const sortedData = dataAttrs.slice().sort((a, b) => {
        const aRank = orderMap.has(a.name) ? orderMap.get(a.name) : Number.MAX_SAFE_INTEGER;
        const bRank = orderMap.has(b.name) ? orderMap.get(b.name) : Number.MAX_SAFE_INTEGER;
        if (aRank !== bRank)
            return aRank - bRank;
        return attributes.indexOf(a) - attributes.indexOf(b);
    });
    return ordered.concat(nonDataAttrs).concat(sortedData);
}
function getPrintWidth(options) {
    return typeof options.printWidth === 'number' && options.printWidth > 0 ? options.printWidth : 80;
}
function getIndentWidth(options) {
    return typeof options.tabWidth === 'number' && Number.isFinite(options.tabWidth) && options.tabWidth > 0
        ? options.tabWidth
        : 2;
}
function getIndentUnit(options) {
    const useTabs = options.useTabs === true;
    const tabWidth = getIndentWidth(options);
    return useTabs ? '\t' : ' '.repeat(tabWidth);
}
function splitMultilineExpression(content) {
    if (!content.includes('\n')) {
        return null;
    }
    const lines = (0, template_format_core_2.trimSurroundingBlankLines)(content.replace(/[ \t]+$/gm, '').split('\n'));
    if (lines.length <= 1) {
        return null;
    }
    return (0, template_format_core_2.stripCommonIndent)(lines, 1);
}
function isStructuralCloseLine(line) {
    const trimmed = line.trim();
    return trimmed.length > 0 && /^[\]})'"`]+$/.test(trimmed);
}
function formatMultilineParamRest(rest, options) {
    if (rest.length === 0) {
        return [];
    }
    const lastLine = rest[rest.length - 1];
    const hasTrailingCloseLine = isStructuralCloseLine(lastLine);
    const bodyLines = hasTrailingCloseLine ? rest.slice(0, -1) : rest;
    const indentUnit = getIndentUnit(options);
    const normalizedBodyLines = (0, template_format_core_2.stripCommonIndent)(bodyLines).map((line) => line.trim() === '' ? '' : `${indentUnit}${line}`);
    const normalizedCloseLine = hasTrailingCloseLine ? lastLine.trim() : null;
    return normalizedCloseLine ? [...normalizedBodyLines, normalizedCloseLine] : normalizedBodyLines;
}
function getEstimatedIndentLength(path, options) {
    let depth = 0;
    for (let ancestorDepth = 0;; ancestorDepth += 1) {
        const ancestor = path.getParentNode(ancestorDepth);
        if (!ancestor) {
            break;
        }
        if (ancestor.type === 'ElementNode' || ancestor.type === 'BlockStatement') {
            depth += 1;
        }
    }
    return depth * getIndentWidth(options);
}
function buildAttributeDocs(attributes, options) {
    const docs = [];
    attributes.forEach((attr) => {
        docs.push(printAttribute(attr, options));
    });
    return docs;
}
function shouldBreakAttribute(attr) {
    if (isRawAttribute(attr)) {
        return /\n/.test(attr.raw);
    }
    if (!isPlainAttribute(attr)) {
        return true;
    }
    if (!attr.value) {
        return false;
    }
    if (attr.value.parts.some((part) => part.type === 'BlockStatement' || part.type === 'CommentStatement')) {
        return true;
    }
    const hasNewlineText = attr.value.parts.some((part) => part.type === 'TextNode' && /\n/.test(part.value));
    if (hasNewlineText) {
        return true;
    }
    return false;
}
function printElement(path, options, print) {
    const node = path.getValue();
    const sortedAttributes = sortAttributes(node.attributes, options);
    const attrsDocs = buildAttributeDocs(sortedAttributes, options);
    const breakAttrs = sortedAttributes.some((attr) => shouldBreakAttribute(attr)) || attrsDocs.some(docHasHardline);
    const parentNode = path.getParentNode();
    const grandParentNode = path.getParentNode(1);
    const ancestors = [
        parentNode,
        grandParentNode,
    ];
    const currentIndentLength = getEstimatedIndentLength(path, options);
    const openTag = concat(['<', node.tag]);
    let attributesDoc = '';
    if (sortedAttributes.length > 0) {
        if (breakAttrs) {
            attributesDoc = concat([
                indent(concat([hardline, join(hardline, attrsDocs)])),
                hardline,
            ]);
        }
        else {
            attributesDoc = node.selfClosing
                ? concat([indent(concat([line, join(line, attrsDocs)])), softline])
                : concat([
                    group(indent(concat([line, join(line, attrsDocs)]))),
                    softline,
                ]);
        }
    }
    const closing = node.selfClosing ? ifBreak('/>', ' />') : '>';
    const tagGroupId = Symbol('tag');
    const openDoc = group(concat([openTag, attributesDoc, closing]), { id: tagGroupId });
    if (node.selfClosing) {
        return openDoc;
    }
    const childrenDocs = [];
    path.each((childPath) => {
        childrenDocs.push(print(childPath));
    }, 'children');
    const closeDoc = concat(['</', node.tag, '>']);
    if (shouldPreserveRawTextElement(node)) {
        return concat([openDoc, node.children[0].value, closeDoc]);
    }
    const singleChild = node.children.length === 1 ? node.children[0] : null;
    if (singleChild?.type === 'TextNode' &&
        getEmbeddedRawTextParser(node, singleChild, options) &&
        childrenDocs.length === 1) {
        return concat([openDoc, indent(concat([hardline, childrenDocs[0]])), hardline, closeDoc]);
    }
    const singleChildIsMustache = singleChild?.type === 'MustacheStatement';
    const mustacheInsideBlock = singleChildIsMustache && ancestors.some((ancestor) => ancestor?.type === 'BlockStatement');
    const openTagFitsInline = !breakAttrs && currentIndentLength + getInlineOpenTagLength(node, sortedAttributes, options) <= getPrintWidth(options);
    const simpleInlineChildren = node.children.length > 0 &&
        node.children.every((child) => (child.type === 'TextNode' && !child.verbatim && !child.blankLines) || child.type === 'MustacheStatement');
    const singleChildCanInline = node.children.length === 1 &&
        childrenDocs.length === 1 &&
        singleChild?.type !== 'ElementNode' &&
        singleChild?.type !== 'PartialStatement' &&
        !docBreaks(childrenDocs[0]) &&
        !mustacheInsideBlock &&
        openTagFitsInline &&
        currentIndentLength + getSingleInlineElementLength(node, sortedAttributes, singleChild, options) <=
            getPrintWidth(options);
    if (singleChildCanInline) {
        return concat([openDoc, childrenDocs[0], closeDoc]);
    }
    const singleTextLikeChildCanUseInlineTag = node.children.length === 1 &&
        childrenDocs.length === 1 &&
        singleChild?.type !== 'ElementNode' &&
        !docBreaks(childrenDocs[0]) &&
        !mustacheInsideBlock &&
        openTagFitsInline;
    if (singleTextLikeChildCanUseInlineTag) {
        return concat([openDoc, indent(concat([hardline, childrenDocs[0]])), hardline, closeDoc]);
    }
    const canInlineSimpleChildren = simpleInlineChildren &&
        !childrenDocs.some(docBreaks) &&
        !mustacheInsideBlock &&
        openTagFitsInline &&
        currentIndentLength + getSimpleInlineElementLength(node, sortedAttributes, options) <= getPrintWidth(options);
    if (canInlineSimpleChildren) {
        return concat([openDoc, joinInlineChildren(node.children, childrenDocs), closeDoc]);
    }
    if (shouldPreserveSimpleInlineText(node, childrenDocs, mustacheInsideBlock)) {
        return stringifySimpleInlineElement(node, sortedAttributes, options);
    }
    const canInlineMixedChildren = isInlineContentTag(node.tag) &&
        node.children.length > 0 &&
        node.children.every(isInlineContentChild) &&
        !childrenDocs.some(docBreaks) &&
        !mustacheInsideBlock &&
        openTagFitsInline &&
        currentIndentLength + getSimpleInlineElementLength(node, sortedAttributes, options) <= getPrintWidth(options);
    if (canInlineMixedChildren) {
        return concat([openDoc, joinInlineChildren(node.children, childrenDocs), closeDoc]);
    }
    const inner = childrenDocs.length > 0
        ? concat([indent(concat([hardline, joinExpandedChildren(node.children, childrenDocs)])), hardline])
        : '';
    const expandedDoc = concat([openDoc, inner, closeDoc]);
    return expandedDoc;
}
function printAttribute(attr, options) {
    if (isRawAttribute(attr)) {
        return attr.raw;
    }
    if (!isPlainAttribute(attr)) {
        if (attr.block.type === 'BlockStatement') {
            return printAttributeBlock(attr.block);
        }
        if (attr.block.type === 'CommentStatement') {
            return printCommentStatement(attr.block, options);
        }
        return stringifyNode(attr.block);
    }
    if (typeof attr.value === 'undefined' || attr.value === null) {
        return attr.name;
    }
    const valueString = stringifyAttributeValue(attr.value);
    const quote = chooseAttributeQuote(valueString, options);
    if (attr.name === 'class' && hasHandlebarsBlock(valueString)) {
        if (classValueHasGluedBlock(attr.value)) {
            const compactValue = stringifyCompactClassValue(attr.value);
            const compactQuote = chooseAttributeQuote(compactValue, options);
            return concat([attr.name, '=', compactQuote, escapeAttributeValue(compactValue, compactQuote), compactQuote]);
        }
        const classLines = formatClassValue(valueString, options);
        return concat([
            'class=',
            quote,
            indent(concat([hardline, join(hardline, classLines)])),
            hardline,
            quote,
        ]);
    }
    if (attr.name === 'class' && shouldExpandStaticClassValue(valueString, options)) {
        const classLines = formatStaticClassValue(valueString);
        return concat([
            'class=',
            quote,
            indent(concat([hardline, join(hardline, classLines)])),
            hardline,
            quote,
        ]);
    }
    if (attr.name === 'class' && hasHandlebarsBlock(valueString)) {
        const lines = formatHandlebarsBlockValue(valueString, options);
        return concat([
            attr.name,
            '=',
            quote,
            indent(concat([hardline, join(hardline, lines)])),
            hardline,
            quote,
        ]);
    }
    if (valueString.includes('\n')) {
        const lines = formatMultilineAttributeValue(valueString);
        return concat([
            attr.name,
            '=',
            quote,
            indent(concat([hardline, join(hardline, lines)])),
            hardline,
            quote,
        ]);
    }
    return concat([attr.name, '=', quote, escapeAttributeValue(valueString, quote), quote]);
}
function printAttributeBlock(block) {
    const open = printBlockOpen(block);
    const bodyLines = formatAttributeBlockBody(stringifyNode(block.program));
    const body = bodyLines.length > 0 ? concat([indent(concat([hardline, join(hardline, bodyLines)])), hardline]) : hardline;
    const inverseParts = [];
    (block.inverseChain ?? []).forEach((branch) => {
        const branchLines = formatAttributeBlockBody(stringifyNode(branch.program));
        const branchBody = branchLines.length > 0 ? concat([indent(concat([hardline, join(hardline, branchLines)])), hardline]) : hardline;
        inverseParts.push(concat([printElseBranchOpen(branch), branchBody]));
    });
    if (block.inverse.body.length > 0) {
        const inverseLines = formatAttributeBlockBody(stringifyNode(block.inverse));
        const elseTag = buildTemplateTag(templateDialect.getElseKeyword());
        inverseParts.push(concat([elseTag, indent(concat([hardline, join(hardline, inverseLines)])), hardline]));
    }
    const inverse = inverseParts.length > 0 ? concat(inverseParts) : '';
    const close = printBlockClose(block);
    return concat([open, body, inverse, close]);
}
function stringifyAttributeValue(value) {
    return value.parts.map((part) => stringifyNode(part)).join('');
}
function getLastClassToken(value) {
    return value.match(/\S+\s*$/)?.[0].trim() ?? '';
}
function classTokenEndsWithContinuation(token) {
    return /[-_:]$/.test(token);
}
function blockProgramToCompactClassValue(program) {
    const parts = [];
    let previousNode = null;
    program.body.forEach((child) => {
        const node = child;
        const value = stringifyCompactClassNode(node);
        if (!value) {
            return;
        }
        if (previousNode && shouldInsertCompactClassSeparator(previousNode, node)) {
            parts.push(' ');
        }
        parts.push(value);
        previousNode = node;
    });
    return parts.join('');
}
function stringifyCompactClassNode(node) {
    switch (node.type) {
        case 'TextNode':
            return (0, template_format_core_2.normalizeInlineText)(node.value);
        case 'MustacheStatement':
            return stringifyMustache(node);
        case 'DecoratorStatement':
            return stringifyDecorator(node);
        case 'PartialStatement': {
            const partial = node;
            return buildTemplateTag(`${templateDialect.getPartialPrefix()}${buildExpression(partial)}`, getTrimOpen(partial), getTrimClose(partial));
        }
        case 'BlockStatement':
            return stringifyCompactClassBlock(node);
        case 'CommentStatement':
            return stringifyNode(node);
        default:
            return stringifyNode(node);
    }
}
function shouldInsertCompactClassSeparator(left, right) {
    if (left.type === 'TextNode') {
        const trailingWhitespace = left.trailingWhitespace;
        if (trailingWhitespace && !hasLineBreak(trailingWhitespace)) {
            return true;
        }
    }
    if (right.type === 'TextNode') {
        const leadingWhitespace = right.leadingWhitespace;
        if (leadingWhitespace && !hasLineBreak(leadingWhitespace)) {
            return true;
        }
    }
    return false;
}
function stringifyCompactClassBlock(block) {
    const prefix = getBlockPrefix(block);
    const printedPrefix = getPrintedBlockPrefix(prefix);
    const expression = buildExpression(block);
    const open = buildTemplateTag(`${printedPrefix}${expression}${getTrimClosePadding(block, expression)}`, getTrimOpen(block), getTrimClose(block));
    const program = blockProgramToCompactClassValue(block.program);
    const inverseChain = (block.inverseChain ?? [])
        .map((branch) => {
        const branchExpression = buildExpression(branch);
        const openBranch = buildTemplateTag(`${templateDialect.getElseKeyword()} ${branchExpression}${getTrimClosePadding(branch, branchExpression)}`, getTrimOpen(branch), getTrimClose(branch));
        return `${openBranch}${blockProgramToCompactClassValue(branch.program)}`;
    })
        .join('');
    const inverse = block.inverse.body.length > 0
        ? `${buildTemplateTag(templateDialect.getElseKeyword(), block.inverseTrimOpen ? '~' : '', block.inverseTrimClose ? '~' : '')}${blockProgramToCompactClassValue(block.inverse)}`
        : '';
    const close = buildTemplateTag(templateDialect.getBlockClosePrefix(block.path), block.closeTrimOpen ? '~' : '', block.closeTrimClose ? '~' : '');
    return `${open}${program}${inverseChain}${inverse}${close}`;
}
function getClassBlockPrograms(block) {
    return [
        block.program,
        ...((block.inverseChain ?? []).map((branch) => branch.program)),
        block.inverse,
    ];
}
function getFirstRenderableProgramNode(program) {
    return program.body.find((node) => stringifyCompactClassNode(node).length > 0);
}
function getLastRenderableProgramNode(program) {
    return program.body
        .slice()
        .reverse()
        .find((node) => stringifyCompactClassNode(node).length > 0);
}
function hasRenderableProgramContent(program) {
    return Boolean(getFirstRenderableProgramNode(program));
}
function programStartsWithClassSeparator(program) {
    const first = getFirstRenderableProgramNode(program);
    return first?.type === 'TextNode' && Boolean(first.leadingWhitespace);
}
function programEndsWithClassSeparator(program) {
    const last = getLastRenderableProgramNode(program);
    return last?.type === 'TextNode' && Boolean(last.trailingWhitespace);
}
function blockStartsWithClassSeparator(block) {
    const programs = getClassBlockPrograms(block).filter(hasRenderableProgramContent);
    return programs.length > 0 && programs.every(programStartsWithClassSeparator);
}
function blockEndsWithClassSeparator(block) {
    const programs = getClassBlockPrograms(block).filter(hasRenderableProgramContent);
    return programs.length > 0 && programs.every(programEndsWithClassSeparator);
}
function shouldGlueTextToFollowingClassBlock(text, block) {
    if (blockStartsWithClassSeparator(block)) {
        return false;
    }
    return !/\s$/.test(text) || classTokenEndsWithContinuation(getLastClassToken(text));
}
function shouldGlueClassBlockToFollowingText(block, text) {
    if (blockEndsWithClassSeparator(block)) {
        return false;
    }
    return !/^\s/.test(text);
}
function programHasGluedClassBlock(program) {
    const body = program.body;
    return body.some((node, index) => {
        if (node.type !== 'BlockStatement') {
            return false;
        }
        const previous = body[index - 1];
        const next = body[index + 1];
        const block = node;
        const isGluedToSibling = (previous?.type === 'TextNode' && shouldGlueTextToFollowingClassBlock(previous.value, block)) ||
            (next?.type === 'TextNode' && shouldGlueClassBlockToFollowingText(block, next.value));
        return isGluedToSibling || classBlockHasNestedGluedBlock(block);
    });
}
function classBlockHasNestedGluedBlock(block) {
    return getClassBlockPrograms(block).some(programHasGluedClassBlock);
}
function classValueHasGluedBlock(value) {
    return value.parts.some((part, index, parts) => {
        if (part.type !== 'BlockStatement') {
            return false;
        }
        const previous = parts[index - 1];
        const next = parts[index + 1];
        return ((previous?.type === 'TextNode' &&
            shouldGlueTextToFollowingClassBlock(previous.value, part)) ||
            (next?.type === 'TextNode' &&
                shouldGlueClassBlockToFollowingText(part, next.value)) ||
            classBlockHasNestedGluedBlock(part));
    });
}
function stringifyCompactClassValue(value) {
    const pieces = [];
    value.parts.forEach((part, index, parts) => {
        const previous = parts[index - 1];
        const next = parts[index + 1];
        if (part.type === 'TextNode') {
            let text = part.value.replace(/\s+/g, ' ');
            if (index === 0) {
                text = text.trimStart();
            }
            if (index === parts.length - 1) {
                text = text.trimEnd();
            }
            if (next?.type === 'BlockStatement' &&
                shouldGlueTextToFollowingClassBlock(part.value, next)) {
                text = text.trimEnd();
            }
            if (previous?.type === 'BlockStatement' &&
                shouldGlueClassBlockToFollowingText(previous, part.value)) {
                text = text.trimStart();
            }
            pieces.push(text);
            return;
        }
        if (part.type === 'BlockStatement') {
            pieces.push(stringifyCompactClassBlock(part));
            return;
        }
        pieces.push(stringifyCompactClassNode(part));
    });
    return pieces.join('').trim();
}
function stringifyAttribute(attr, options) {
    if (isRawAttribute(attr)) {
        return attr.raw;
    }
    if (!isPlainAttribute(attr)) {
        return stringifyNode(attr.block);
    }
    if (!attr.value) {
        return attr.name;
    }
    const value = stringifyAttributeValue(attr.value);
    const quote = chooseAttributeQuote(value, options);
    return `${attr.name}=${quote}${escapeAttributeValue(value, quote)}${quote}`;
}
function getInlineOpenTagLength(node, attributes, options) {
    const attrs = attributes.map((attr) => stringifyAttribute(attr, options)).join(' ');
    const open = attrs ? `<${node.tag} ${attrs}` : `<${node.tag}`;
    const close = node.selfClosing ? ' />' : '>';
    return `${open}${close}`.length;
}
function stringifyInlineChild(node, options) {
    switch (node.type) {
        case 'TextNode':
            return (0, template_format_core_2.normalizeInlineText)(node.value);
        case 'MustacheStatement':
            return stringifyMustache(node);
        case 'ElementNode': {
            const element = node;
            const sortOptions = options ?? {};
            const sortedAttributes = sortAttributes(element.attributes, sortOptions);
            if (isInlineContentTag(element.tag) &&
                element.children.length > 0 &&
                element.children.every(isInlineContentChild) &&
                !sortedAttributes.some(shouldBreakAttribute)) {
                return stringifySimpleInlineElement(element, sortedAttributes, sortOptions);
            }
            return stringifyNode(node);
        }
        default:
            return stringifyNode(node);
    }
}
function shouldInsertInlineSeparator(left, right) {
    if (left.type === 'TextNode' && hasInlineBoundaryWhitespace(left.trailingWhitespace)) {
        return !isPunctuationOnlyTextNode(left) || !hasLineBreak(left.trailingWhitespace);
    }
    if (right.type === 'TextNode' && hasInlineBoundaryWhitespace(right.leadingWhitespace)) {
        return !isPunctuationOnlyTextNode(right) || !hasLineBreak(right.leadingWhitespace);
    }
    if (isPunctuationOnlyTextNode(left) || isPunctuationOnlyTextNode(right)) {
        return false;
    }
    return left.type !== 'TextNode' && right.type !== 'TextNode';
}
function shouldAttachExpandedChild(left, right) {
    return Boolean(left) && (isPunctuationOnlyTextNode(left) || isPunctuationOnlyTextNode(right));
}
function joinInlineChildren(nodes, docs) {
    const parts = [];
    docs.forEach((doc, index) => {
        if (index > 0 && shouldInsertInlineSeparator(nodes[index - 1], nodes[index])) {
            parts.push(' ');
        }
        parts.push(doc);
    });
    return concat(parts);
}
function joinExpandedChildren(nodes, docs) {
    const parts = [];
    docs.forEach((doc, index) => {
        if (index > 0 && !shouldAttachExpandedChild(nodes[index - 1], nodes[index])) {
            parts.push(hardline);
        }
        parts.push(doc);
    });
    return concat(parts);
}
function printBlockBody(children, docs, options) {
    if (docs.length === 0) {
        return hardline;
    }
    if (canPrintInlineTextProgram(children, options)) {
        return concat([indent(concat([hardline, stringifyInlineChildren(children, options)])), hardline]);
    }
    return concat([indent(concat([hardline, join(hardline, docs)])), hardline]);
}
function canPrintInlineTextProgram(nodes, options) {
    return (nodes.some((node) => node.type === 'TextNode') &&
        nodes.every((node, index) => isInlineTextProgramChild(node, index, nodes, options)));
}
function isInlineTextProgramChild(node, index, nodes, options) {
    if (node.type === 'MustacheStatement' || node.type === 'PartialStatement' || node.type === 'DecoratorStatement') {
        return true;
    }
    if (node.type === 'CommentStatement') {
        const comment = node;
        return !comment.block && !comment.multiline;
    }
    if (node.type === 'BlockStatement') {
        return canInlineBlock(node, options, 'Program');
    }
    if (node.type !== 'TextNode') {
        return false;
    }
    const text = node;
    if (text.verbatim || text.blankLines || /[\r\n]/.test(text.value)) {
        return false;
    }
    const leadingBreakAllowed = index === 0 || !hasLineBreak(text.leadingWhitespace);
    const trailingBreakAllowed = index === nodes.length - 1 || !hasLineBreak(text.trailingWhitespace);
    return leadingBreakAllowed && trailingBreakAllowed;
}
function stringifyInlineChildren(nodes, options) {
    return nodes.reduce((result, child, index) => {
        const separator = index > 0 && shouldInsertInlineSeparator(nodes[index - 1], child) ? ' ' : '';
        return `${result}${separator}${stringifyInlineChild(child, options)}`;
    }, '');
}
function stringifySimpleInlineElement(node, attributes, options) {
    const attrs = attributes.map((attr) => stringifyAttribute(attr, options)).join(' ');
    const open = attrs ? `<${node.tag} ${attrs}>` : `<${node.tag}>`;
    return `${open}${stringifyInlineChildren(node.children, options)}</${node.tag}>`;
}
function shouldPreserveRawTextElement(node) {
    return (template_format_core_1.whitespaceSensitiveRawTextElements.has(node.tag.toLowerCase()) &&
        node.children.length === 1 &&
        node.children[0].type === 'TextNode' &&
        node.children[0].preserveWhitespace === true);
}
function getSingleInlineElementLength(node, attributes, child, options) {
    return getInlineOpenTagLength(node, attributes, options) + stringifyInlineChild(child, options).length + `</${node.tag}>`.length;
}
function getSimpleInlineElementLength(node, attributes, options) {
    const childrenLength = stringifyInlineChildren(node.children, options).length;
    return getInlineOpenTagLength(node, attributes, options) + childrenLength + `</${node.tag}>`.length;
}
function shouldPreserveSimpleInlineText(node, childrenDocs, mustacheInsideBlock) {
    return (isInlineContentTag(node.tag) &&
        node.children.some((child) => child.type === 'MustacheStatement') &&
        node.children.every((child) => (child.type === 'TextNode' && !child.verbatim && !child.blankLines) ||
            child.type === 'MustacheStatement') &&
        !childrenDocs.some(docBreaks) &&
        !mustacheInsideBlock);
}
function isInlineContentTag(tag) {
    return template_format_core_1.inlineContentElements.has(tag.toLowerCase());
}
function isInlineContentChild(node) {
    if ((node.type === 'TextNode' && !node.verbatim && !node.blankLines) || node.type === 'MustacheStatement') {
        return true;
    }
    return node.type === 'ElementNode' && isInlineContentTag(node.tag);
}
function isSimpleInlineBlockNode(node, options, parentType) {
    switch (node.type) {
        case 'TextNode':
            return !node.verbatim && !node.blankLines;
        case 'MustacheStatement':
        case 'PartialStatement':
            return true;
        case 'BlockStatement':
            return canInlineBlock(node, options, parentType);
        default:
            return false;
    }
}
function canInlineBlock(node, options, parentType) {
    const blockPrefix = getBlockPrefix(node);
    if (blockPrefix !== '#' && blockPrefix !== '^' && blockPrefix !== '#*' && blockPrefix !== '$') {
        return false;
    }
    if (hasOriginalLineBreak(node, options)) {
        return false;
    }
    const programChildren = node.program.body;
    const inverseChildren = node.inverse.body;
    const inverseChainChildren = (node.inverseChain ?? []).flatMap((branch) => branch.program.body);
    const allChildren = [...programChildren, ...inverseChainChildren, ...inverseChildren];
    if (allChildren.length === 0) {
        return false;
    }
    if (parentType === 'ElementNode') {
        return false;
    }
    if (parentType && parentType !== 'Program' && parentType !== 'BlockStatement') {
        return false;
    }
    if (!allChildren.every((child) => isSimpleInlineBlockNode(child, options, 'BlockStatement'))) {
        return false;
    }
    return stringifyNode(node).length <= getPrintWidth(options);
}
function hasOriginalLineBreak(node, options) {
    const range = node.range;
    const originalText = options.originalText;
    return Boolean(range && originalText && /[\r\n]/.test(originalText.slice(range[0], range[1])));
}
function stringifyNode(node) {
    switch (node.type) {
        case 'TextNode':
            return node.value;
        case 'MustacheStatement': {
            const mustache = node;
            return stringifyMustache(mustache);
        }
        case 'DecoratorStatement': {
            return stringifyDecorator(node);
        }
        case 'PartialStatement': {
            const partial = node;
            return buildTemplateTag(`${templateDialect.getPartialPrefix()}${buildExpression(partial)}`, getTrimOpen(partial), getTrimClose(partial));
        }
        case 'CommentStatement': {
            const comment = node;
            if (comment.block || comment.multiline) {
                return templateDialect.getBlockCommentTag(comment.value);
            }
            return templateDialect.getLineCommentTag(comment.value);
        }
        case 'BlockStatement': {
            const block = node;
            const prefix = getBlockPrefix(block);
            const printedPrefix = getPrintedBlockPrefix(prefix);
            const expression = buildExpression(block);
            const open = buildTemplateTag(`${printedPrefix}${expression}${getTrimClosePadding(block, expression)}`, getTrimOpen(block), getTrimClose(block));
            const program = stringifyInlineChildren(block.program.body);
            const inverseChain = (block.inverseChain ?? [])
                .map((branch) => {
                const branchExpression = buildExpression(branch);
                const openBranch = buildTemplateTag(`${templateDialect.getElseKeyword()} ${branchExpression}${getTrimClosePadding(branch, branchExpression)}`, getTrimOpen(branch), getTrimClose(branch));
                return `${openBranch}${stringifyInlineChildren(branch.program.body)}`;
            })
                .join('');
            const inverse = block.inverse.body.length > 0
                ? `${buildTemplateTag(templateDialect.getElseKeyword(), block.inverseTrimOpen ? '~' : '', block.inverseTrimClose ? '~' : '')}${stringifyInlineChildren(block.inverse.body)}`
                : '';
            const close = buildTemplateTag(templateDialect.getBlockClosePrefix(block.path), block.closeTrimOpen ? '~' : '', block.closeTrimClose ? '~' : '');
            return `${open}${program}${inverseChain}${inverse}${close}`;
        }
        case 'ElementNode': {
            const element = node;
            const attrs = element.attributes.map((attr) => stringifyAttribute(attr)).join(' ');
            const open = attrs ? `<${element.tag} ${attrs}${element.selfClosing ? ' />' : '>'}` : `<${element.tag}${element.selfClosing ? ' />' : '>'}`;
            if (element.selfClosing) {
                return open;
            }
            const children = element.children.map((child) => stringifyNode(child)).join('');
            return `${open}${children}</${element.tag}>`;
        }
        case 'Program':
            return node.body.map((child) => stringifyNode(child)).join('');
        case 'UnmatchedNode':
            return node.raw;
        default:
            return '';
    }
}
function formatClassValue(value, options) {
    const tokens = tokenizeClass(value);
    const lines = [];
    let depth = 0;
    tokens.forEach((token) => {
        if (token.startsWith('{{/')) {
            depth = Math.max(depth - 1, 0);
            lines.push(indentWithDepth(token, depth, options));
            return;
        }
        if (token.startsWith('{{#') || token.startsWith('{{^')) {
            lines.push(indentWithDepth(token, depth, options));
            depth += 1;
            return;
        }
        if (token.startsWith('{{else')) {
            depth = Math.max(depth - 1, 0);
            lines.push(indentWithDepth(token, depth, options));
            depth += 1;
            return;
        }
        lines.push(indentWithDepth(token, depth, options));
    });
    return lines;
}
function indentWithDepth(content, depth, options) {
    const prefix = depth > 0 ? getIndentUnit(options).repeat(depth) : '';
    return concat([prefix, content]);
}
function tokenizeClass(value) {
    const tokens = [];
    const mustacheRegex = /{{[^}]+}}/g;
    let lastIndex = 0;
    let match;
    while ((match = mustacheRegex.exec(value)) !== null) {
        const before = value.slice(lastIndex, match.index);
        before.split(/\s+/).filter(Boolean).forEach((word) => tokens.push(word));
        tokens.push(match[0].trim());
        lastIndex = match.index + match[0].length;
    }
    const remaining = value.slice(lastIndex);
    remaining.split(/\s+/).filter(Boolean).forEach((word) => tokens.push(word));
    return mergeClassTokenFragments(tokens);
}
function isSimpleMustacheToken(token) {
    return (token.startsWith('{{') &&
        !token.startsWith('{{#') &&
        !token.startsWith('{{/') &&
        !token.startsWith('{{else'));
}
function shouldGlueClassTokens(left, right) {
    return ((isSimpleMustacheToken(right) && /[-_:]$/.test(left)) ||
        (isSimpleMustacheToken(left) && /^[-_:]/.test(right)));
}
function mergeClassTokenFragments(tokens) {
    const merged = [];
    tokens.forEach((token) => {
        const previous = merged[merged.length - 1];
        if (previous && shouldGlueClassTokens(previous, token)) {
            merged[merged.length - 1] = `${previous}${token}`;
            return;
        }
        merged.push(token);
    });
    return merged;
}
function formatStaticClassValue(value) {
    return value
        .split(/\s+/)
        .filter(Boolean)
        .map((token) => token);
}
function chooseAttributeQuote(value, options) {
    const preferSingleQuote = options?.singleQuote === true;
    if (preferSingleQuote && !value.includes("'")) {
        return "'";
    }
    if (value.includes('"') && !value.includes("'")) {
        return "'";
    }
    return '"';
}
function escapeAttributeValue(value, quote) {
    if (quote === '"') {
        return value.replace(/"/g, '&quot;');
    }
    return value.replace(/'/g, '&#39;');
}
function shouldExpandStaticClassValue(value, options) {
    const tokens = value.split(/\s+/).filter(Boolean);
    if (tokens.length < 2) {
        return false;
    }
    return `class="${value.trim()}"`.length > getPrintWidth(options);
}
function stringifyMustache(node) {
    const { open, close } = getTemplateTagDelimiters(node.triple);
    const content = buildExpression(node);
    return `${open}${getTrimOpen(node)}${getMustacheOpenPadding(node, content)}${content}${getMustacheClosePadding(node, content)}${getTrimClose(node)}${close}`;
}
function stringifyDecorator(node) {
    const content = buildExpression(node);
    return buildTemplateTag(`${templateDialect.getDecoratorPrefix()}${content}${getTrimClosePadding(node, content)}`, getTrimOpen(node), getTrimClose(node));
}
function shouldPrintCallableMultiline(node, content, options, allowHashWrap) {
    const expressionPartCount = node.hash.length + node.params.length;
    const canWrapPlainParams = !node.params.some(shouldKeepParamInline);
    return (expressionPartCount > 1 &&
        ((allowHashWrap && node.hash.length > 0) || (canWrapPlainParams && content.length > getPrintWidth(options))));
}
function buildCallableParamsDocs(node) {
    const paramsDocs = [];
    node.params.forEach((param) => paramsDocs.push(param));
    node.hash.forEach((pair) => paramsDocs.push(formatHash(pair)));
    return paramsDocs;
}
function printCallableStatement(node, config) {
    if (config.multiline) {
        return group(concat([
            config.open,
            getTrimOpen(node),
            config.multilineHead,
            indent(concat([hardline, join(hardline, buildCallableParamsDocs(node))])),
            hardline,
            getTrimClose(node),
            config.close,
        ]));
    }
    return concat([
        config.open,
        getTrimOpen(node),
        config.openPadding ?? '',
        config.inlineContent,
        config.closePadding,
        getTrimClose(node),
        config.close,
    ]);
}
function printMustache(node, options) {
    const content = buildExpression(node);
    const { open, close } = getTemplateTagDelimiters(node.triple);
    return printCallableStatement(node, {
        open,
        close,
        inlineContent: content,
        multilineHead: node.path,
        openPadding: getMustacheOpenPadding(node, content),
        closePadding: getMustacheClosePadding(node, content),
        multiline: !node.triple && shouldPrintCallableMultiline(node, content, options, true),
    });
}
function printDecorator(node, options) {
    const content = buildExpression(node);
    const { open, close } = getTemplateTagDelimiters(false);
    const decoratorPrefix = templateDialect.getDecoratorPrefix();
    return printCallableStatement(node, {
        open,
        close,
        inlineContent: `${decoratorPrefix}${content}`,
        multilineHead: `${decoratorPrefix}${node.path}`,
        closePadding: getTrimClosePadding(node, content),
        multiline: shouldPrintCallableMultiline(node, content, options, false),
    });
}
function printBlock(path, options, print) {
    const node = path.getValue();
    const parentNode = path.getParentNode();
    if (canInlineBlock(node, options, parentNode?.type)) {
        return stringifyNode(node);
    }
    const open = printBlockOpen(node);
    const bodyDocs = [];
    path.call((programPath) => {
        programPath.each((childPath) => {
            const childNode = childPath.getValue();
            if (childNode.type === 'TextNode' && childNode.blankLines && getMaxEmptyLines(options) === 0) {
                return;
            }
            const doc = print(childPath);
            if (doc === null) {
                return;
            }
            bodyDocs.push(doc);
        }, 'body');
    }, 'program');
    const body = printBlockBody(node.program.body, bodyDocs, options);
    const inverseParts = [];
    (node.inverseChain ?? []).forEach((branch, index) => {
        const branchDocs = [];
        path.call((branchProgramPath) => {
            branchProgramPath.each((childPath) => {
                const childNode = childPath.getValue();
                if (childNode.type === 'TextNode' && childNode.blankLines && getMaxEmptyLines(options) === 0) {
                    return;
                }
                const doc = print(childPath);
                if (doc === null) {
                    return;
                }
                branchDocs.push(doc);
            }, 'body');
        }, 'inverseChain', index, 'program');
        const branchBody = printBlockBody(branch.program.body, branchDocs, options);
        inverseParts.push(concat([printElseBranchOpen(branch), branchBody]));
    });
    if (node.inverse.body.length > 0) {
        const inverseDocs = [];
        path.call((inversePath) => {
            inversePath.each((childPath) => {
                const childNode = childPath.getValue();
                if (childNode.type === 'TextNode' && childNode.blankLines && getMaxEmptyLines(options) === 0) {
                    return;
                }
                const doc = print(childPath);
                if (doc === null) {
                    return;
                }
                inverseDocs.push(doc);
            }, 'body');
        }, 'inverse');
        inverseParts.push(concat([printFinalElseOpen(node), printBlockBody(node.inverse.body, inverseDocs, options)]));
    }
    const inverse = inverseParts.length > 0 ? concat(inverseParts) : '';
    const close = printBlockClose(node);
    return concat([open, body, inverse, close]);
}
function printBlockOpen(node) {
    const expression = buildExpression(node);
    const prefix = getBlockPrefix(node);
    const printedPrefix = getPrintedBlockPrefix(prefix);
    return printExpressionTag(printedPrefix, expression, node);
}
function getPrintedBlockPrefix(prefix) {
    return templateDialect.getPrintedBlockPrefix(prefix);
}
function printExpressionTag(head, expression, node) {
    const { open, close } = getTemplateTagDelimiters(false);
    const multilineExpression = splitMultilineExpression(expression);
    if (multilineExpression) {
        return concat([
            open,
            getTrimOpen(node),
            head,
            multilineExpression[0],
            indent(concat([hardline, join(hardline, multilineExpression.slice(1))])),
            hardline,
            getTrimClose(node),
            close,
        ]);
    }
    return concat([
        open,
        getTrimOpen(node),
        head,
        expression,
        getTrimClosePadding(node, expression),
        getTrimClose(node),
        close,
    ]);
}
function printFinalElseOpen(node) {
    return buildTemplateTag(templateDialect.getElseKeyword(), node.inverseTrimOpen ? '~' : '', node.inverseTrimClose ? '~' : '');
}
function printElseBranchOpen(node) {
    const expression = buildExpression(node);
    return printExpressionTag(`${templateDialect.getElseKeyword()} `, expression, node);
}
function printBlockClose(node) {
    return buildTemplateTag(templateDialect.getBlockClosePrefix(node.path), node.closeTrimOpen ? '~' : '', node.closeTrimClose ? '~' : '');
}
function printPartial(node, options) {
    const name = node.path;
    const { open: tagOpen, close: tagClose } = getTemplateTagDelimiters(false);
    const open = concat([tagOpen, getTrimOpen(node), templateDialect.getPartialPrefix()]);
    const close = concat([getTrimClose(node), tagClose]);
    if (node.params.length === 0 && node.hash.length === 0) {
        return concat([open, name, close]);
    }
    const paramsDocs = [];
    node.params.forEach((param) => paramsDocs.push(formatPartialParam(param, options)));
    node.hash.forEach((pair) => paramsDocs.push(formatPartialParam(formatHash(pair), options)));
    return group(concat([
        open,
        name,
        indent(concat([hardline, join(hardline, paramsDocs)])),
        hardline,
        close,
    ]));
}
function formatPartialParam(param, options) {
    if (!param.includes('\n')) {
        return param;
    }
    const lines = (0, template_format_core_2.trimSurroundingBlankLines)(param.split('\n'));
    if (lines.length === 0) {
        return '';
    }
    const [firstLine, ...rest] = lines;
    if (rest.length === 0) {
        return firstLine;
    }
    const normalizedRest = formatMultilineParamRest(rest, options);
    return concat([firstLine, hardline, join(hardline, normalizedRest)]);
}
function formatMultilineComment(content, options, inlineMarkers = false) {
    const hasStandaloneCloseMarker = inlineMarkers && /\n[ \t]*$/.test(content);
    const lines = (0, template_format_core_2.trimSurroundingBlankLines)(content.replace(/[ \t]+$/gm, '').split('\n'));
    const shouldInlineMarkup = !inlineMarkers && shouldFormatCommentAsInlineMarkup(lines);
    const markers = templateDialect.getBlockCommentMarkers();
    if (lines.length === 0) {
        return inlineMarkers ? markers.emptyInline : markers.emptyBlock;
    }
    if (inlineMarkers || shouldInlineMarkup) {
        const strippedLines = inlineMarkers ? (0, template_format_core_2.stripCommonIndent)(lines, 1) : (0, template_format_core_2.stripCommonIndent)(lines);
        const [firstLine, ...restLines] = strippedLines;
        const first = firstLine.trimStart();
        if (restLines.length === 0) {
            return hasStandaloneCloseMarker
                ? concat([markers.inlineOpen, first, hardline, markers.blockClose])
                : concat([markers.inlineOpen, first, markers.inlineClose]);
        }
        const normalizedRest = normalizeInlineCommentLines(restLines, options).map((line) => {
            if (!hasStandaloneCloseMarker || line.trim() === '') {
                return line;
            }
            return `${getIndentUnit(options)}${line}`;
        });
        const lastLine = normalizedRest[normalizedRest.length - 1];
        const leadingLines = normalizedRest.slice(0, -1);
        if (hasStandaloneCloseMarker) {
            return concat([
                markers.inlineOpen,
                first,
                hardline,
                leadingLines.length > 0 ? concat([join(hardline, leadingLines), hardline]) : '',
                lastLine,
                hardline,
                markers.blockClose,
            ]);
        }
        return concat([
            markers.inlineOpen,
            first,
            hardline,
            leadingLines.length > 0 ? concat([join(hardline, leadingLines), hardline]) : '',
            lastLine,
            markers.inlineClose,
        ]);
    }
    const normalizedLines = (0, template_format_core_2.stripCommonIndent)(lines).map((line) => {
        if (line.trim() === '') {
            return '';
        }
        return `${getIndentUnit(options)}${line}`;
    });
    const body = join(hardline, normalizedLines);
    return concat([
        markers.blockOpen,
        hardline,
        body,
        hardline,
        markers.blockClose,
    ]);
}
function shouldFormatCommentAsInlineMarkup(lines) {
    const nonEmptyLines = lines.filter((line) => line.trim() !== '');
    if (nonEmptyLines.length < 2) {
        return false;
    }
    const firstLine = nonEmptyLines[0].trim();
    const lastLine = nonEmptyLines[nonEmptyLines.length - 1].trim();
    return /^<[\w:-]+$/.test(firstLine) && (/\/>$/.test(lastLine) || /^<\/[\w:-]+>$/.test(lastLine));
}
function normalizeInlineCommentLines(lines, options) {
    const indentUnit = getIndentUnit(options);
    const nonEmptyLines = lines.filter((line) => line.trim() !== '');
    if (nonEmptyLines.length === 0) {
        return lines.map(() => '');
    }
    const baseIndent = nonEmptyLines.reduce((min, line) => {
        const indentLength = (line.match(/^[ \t]*/) || [''])[0].length;
        return Math.min(min, indentLength);
    }, Number.MAX_SAFE_INTEGER);
    const relativeIndentLengths = Array.from(new Set(nonEmptyLines.map((line) => {
        const indentLength = (line.match(/^[ \t]*/) || [''])[0].length;
        return Math.max(indentLength - baseIndent, 0);
    }))).sort((a, b) => a - b);
    const indentRank = new Map(relativeIndentLengths.map((length, index) => [length, index]));
    return lines.map((line) => {
        if (line.trim() === '') {
            return '';
        }
        const indentLength = (line.match(/^[ \t]*/) || [''])[0].length;
        const relativeIndent = Math.max(indentLength - baseIndent, 0);
        const level = indentRank.get(relativeIndent) ?? 0;
        const content = line.slice(indentLength).replace(/[ \t]+$/, '');
        return `${indentUnit.repeat(level)}${content}`;
    });
}
function buildExpression(node) {
    const pieces = [];
    if (node.path) {
        pieces.push(node.path);
    }
    if (node.params.length > 0) {
        pieces.push(...node.params);
    }
    if (node.hash.length > 0) {
        pieces.push(...node.hash.map((pair) => formatHash(pair)));
    }
    if (node.blockParams && node.blockParams.length > 0) {
        pieces.push('as', `|${node.blockParams.join(' ')}|`);
    }
    return pieces.join(' ');
}
function formatHash(pair) {
    return `${pair.key}=${pair.value}`;
}
function formatMultilineAttributeValue(value) {
    const lines = value.split('\n');
    while (lines.length > 0 && lines[0].trim() === '') {
        lines.shift();
    }
    while (lines.length > 0 && lines[lines.length - 1].trim() === '') {
        lines.pop();
    }
    const commonIndent = lines.reduce((min, line) => {
        if (line.trim() === '')
            return min;
        const indentLength = (line.match(/^[ \t]*/) || [''])[0].length;
        return Math.min(min, indentLength);
    }, Number.MAX_SAFE_INTEGER);
    const normalizedIndent = Number.isFinite(commonIndent) ? commonIndent : 0;
    return lines.map((line) => {
        const indentLength = (line.match(/^[ \t]*/) || [''])[0].length;
        const trimmedLine = line.slice(Math.min(indentLength, normalizedIndent));
        return trimmedLine.replace(/[ \t]+$/, '');
    });
}
function hasHandlebarsBlock(value) {
    return /{{[#/^]/.test(value) && /{{\//.test(value);
}
function formatHandlebarsBlockValue(value, options) {
    const tokens = [];
    const mustacheRegex = /{{[^}]+}}/g;
    let lastIndex = 0;
    let match;
    while ((match = mustacheRegex.exec(value)) !== null) {
        const before = value.slice(lastIndex, match.index).trim();
        if (before) {
            tokens.push(before);
        }
        tokens.push(match[0].trim());
        lastIndex = match.index + match[0].length;
    }
    const remaining = value.slice(lastIndex).trim();
    if (remaining) {
        tokens.push(remaining);
    }
    const lines = [];
    let depth = 0;
    tokens.forEach((token) => {
        const isClosing = token.startsWith('{{/');
        const isElse = token.startsWith('{{else');
        const isBlockOpen = token.startsWith('{{#') || token.startsWith('{{^');
        if (isClosing || isElse) {
            depth = Math.max(depth - 1, 0);
        }
        lines.push(indentWithDepth(token, depth, options));
        if (isBlockOpen || isElse) {
            depth += 1;
        }
    });
    return lines;
}
function formatAttributeBlockBody(value) {
    return value
        .split('\n')
        .map((line) => line.replace(/[ \t]+$/, ''))
        .filter((line, index, lines) => {
        if (line.trim() !== '') {
            return true;
        }
        return index > 0 && index < lines.length - 1;
    })
        .map((line) => line.trim())
        .filter((line) => line.length > 0);
}
