// template-format-core/dist/template/expression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseTemplateExpression = parseTemplateExpression;
exports.normalizeTemplateExpression = normalizeTemplateExpression;
exports.tokenizeTemplateExpression = tokenizeTemplateExpression;
exports.isTemplateExpressionQuoteStart = isTemplateExpressionQuoteStart;
exports.splitTemplateParams = splitTemplateParams;
exports.formatTemplateHashPair = formatTemplateHashPair;
const whitespace_1 = require("../text/whitespace");
function parseTemplateExpression(content) {
    const normalized = normalizeTemplateExpression(content);
    const { expression, blockParams } = extractBlockParams(normalized);
    const tokens = tokenizeTemplateExpression(expression);
    const path = tokens.shift() || '';
    const { params, hash } = splitTemplateParams(tokens);
    const parsed = {
        path,
        params,
        hash,
    };
    if (blockParams.length > 0) {
        parsed.blockParams = blockParams;
    }
    return parsed;
}
function normalizeTemplateExpression(content) {
    const joined = tokenizeTemplateExpression(content.trim()).join(' ');
    return joined.replace(/\(\s+/g, '(').replace(/\s+\)/g, ')');
}
function tokenizeTemplateExpression(content) {
    const tokens = [];
    let current = '';
    let quote = null;
    let parenDepth = 0;
    let bracketDepth = 0;
    let braceDepth = 0;
    for (let i = 0; i < content.length; i += 1) {
        const char = content[i];
        if (quote) {
            current += char;
            if (char === '\\' && i + 1 < content.length) {
                i += 1;
                current += content[i];
                continue;
            }
            if (char === quote) {
                quote = null;
            }
            continue;
        }
        if ((char === '"' || char === "'" || char === '`') && isTemplateExpressionQuoteStart(content, i)) {
            current += char;
            quote = char;
            continue;
        }
        if (char === '(') {
            parenDepth += 1;
            current += char;
            continue;
        }
        if (char === ')') {
            parenDepth = Math.max(parenDepth - 1, 0);
            current += char;
            continue;
        }
        if (char === '[') {
            bracketDepth += 1;
            current += char;
            continue;
        }
        if (char === ']') {
            bracketDepth = Math.max(bracketDepth - 1, 0);
            current += char;
            continue;
        }
        if (char === '{') {
            braceDepth += 1;
            current += char;
            continue;
        }
        if (char === '}') {
            braceDepth = Math.max(braceDepth - 1, 0);
            current += char;
            continue;
        }
        if (whitespace_1.whitespace.test(char) && parenDepth === 0 && bracketDepth === 0 && braceDepth === 0) {
            if (current) {
                tokens.push(current);
                current = '';
            }
            continue;
        }
        current += char;
    }
    if (current) {
        tokens.push(current);
    }
    return tokens;
}
function isTemplateExpressionQuoteStart(text, position, expressionStart = 0) {
    if (position <= expressionStart) {
        return true;
    }
    const previous = text[position - 1];
    return !previous || whitespace_1.whitespace.test(previous) || /[([{=,:~|]/.test(previous);
}
function splitTemplateParams(tokens) {
    const params = [];
    const hash = [];
    for (let index = 0; index < tokens.length; index += 1) {
        const token = tokens[index];
        const eqIndex = token.indexOf('=');
        const key = eqIndex > 0 ? token.slice(0, eqIndex) : '';
        if (eqIndex > 0 && isHashKey(key)) {
            let value = token.slice(eqIndex + 1);
            if (value === '' && index + 1 < tokens.length) {
                value = tokens[index + 1];
                index += 1;
            }
            hash.push({ key, value });
            continue;
        }
        if (isHashKey(token) && tokens[index + 1] === '=' && index + 2 < tokens.length) {
            hash.push({ key: token, value: tokens[index + 2] });
            index += 2;
            continue;
        }
        params.push(token);
    }
    return { params, hash };
}
function formatTemplateHashPair(pair) {
    return `${pair.key}=${pair.value}`;
}
function extractBlockParams(content) {
    const match = content.match(/\s+as\s+\|([^|]*)\|/);
    if (!match) {
        return { expression: content, blockParams: [] };
    }
    const blockParams = match[1].split(/\s+/).filter(Boolean);
    const expression = content.replace(match[0], '').trim();
    return { expression, blockParams };
}
function isHashKey(value) {
    return /^[A-Za-z_@][A-Za-z0-9_@.:-]*$/.test(value);
}
